### PR TITLE
Refactor AddressSearch to be compatible with Form

### PR DESCRIPTION
**Pull request by kakajann**
Done.

> 4. Make the value prop optional

It was already optional.

> 5. In the input’s onBlur method, call props.onBlur().
> 6. In the input’s onChange method (or equivalent e.g. onTextChange, etc), call props.onChange().

Done.

> 7. Remove the hasError prop

There was no hasError prop.

> 8. Add an optional errorText prop. Defaults to an empty string.
> 9. Update the error message to display if errorText is truth.

Done. 

> 10. Update the inline error display style so it is left aligned with the label and input value. See example image for TextInput

I don't think it's my concern. `AdressSearch` is using `TextInput`

> 11. Make sure that props.ref is attached to the appropriate DOM node. This could involve forwarding the ref to a child component. This is important so we can call ref.focus().

Done.

> 12. Add the input to the test [Form stories](https://github.com/Expensify/App/blob/1860b590114991d1819a5590608d26c32337de56/src/stories/Form.stories.js) and make sure that it works. You can check this by running npm run storybook and testing the component in the example form.

Done.

> 13. Remove any unused code.

Nothing to remove.

### Fixed Issues

$ https://github.com/Expensify/App/issues/7538

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

Nothing to test I guess. It's only refactor.

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [X] Desktop
- [x] iOS
- [x] Android

### Screenshots

Screen Shot 2022-02-11 at 9 59 19 AM
Screen Shot 2022-02-11 at 9 59 07 AM
Screen Shot 2022-02-11 at 9 59 00 AM

### CODE REVIEW

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@kakajann two things -
1. Error from storybook - `props.onChange` isn't a function.
image

2. Point 10 isn't implemented - Error text needs to be aligned with label and input value as show here - https://github.com/Expensify/App/issues/7538#issue-1122427844

**Comment by kakajann** [comment]
> 1. Error from storybook - `props.onChange` isn't a function.
>    image

I'll take a look at this soon.

> 2. Point 10 isn't implemented - Error text needs to be aligned with label and input value as show here - [#7538 (comment)](https://github.com/Expensify/App/issues/7538#issue-1122427844)

As I mentioned above, `AddressSearch` is using `TextInput`. I didn't want to modify it. It should have been done in this [PR](https://github.com/Expensify/App/pull/6412)

Also, this PR covers the error alignment: https://github.com/Expensify/App/pull/7627/files

Do you still want me to modify error alignment?

**Comment by kakajann** [comment]
@rushatgabhane 
> Error from storybook - props.onChange isn't a function

is fixed

File: src/components/AddressSearch.js [hunk 1]
```diff
@@ -32,6 +53,12 @@ const propTypes = {
 };
 
 const defaultProps = {
+    // ...defaultFieldPropTypes,
+    isFormInput: false,
+    inputID: undefined,
+    shouldSaveDraft: false,
+    onBlur: () => {},
+    errorText: '',
     value: '',
```
Comment (luacmartins) [inline src/components/AddressSearch.js:62]: We should update the default value to `undefined`

File: src/components/AddressSearch.js [hunk 2]
```diff
@@ -32,6 +53,12 @@ const propTypes = {
 };
 
 const defaultProps = {
+    // ...defaultFieldPropTypes,
```
Comment (luacmartins) [inline src/components/AddressSearch.js:56]: Remove commented `defaultProps`

File: src/components/AddressSearch.js [hunk 3]
```diff
@@ -111,11 +138,28 @@ const AddressSearch = (props) => {
             }}
             textInputProps={{
                 InputComp: TextInput,
+                ref: (node) => {
+                    if (!props.innerRef) {
+                        return;
+                    }
+
```
Comment (luacmartins) [inline src/components/AddressSearch.js:145]: I think we can remove this check as `_.isFunction` would return false if `props.innerRef` is undefined.

File: src/components/AddressSearch.js [hunk 4]
```diff
@@ -111,11 +138,28 @@ const AddressSearch = (props) => {
...
                 label: props.label,
                 containerStyles: props.containerStyles,
                 errorText: props.errorText,
                 value: props.value,
-                onChangeText: (text) => {
+                isFormInput: props.isFormInput,
+                inputID: props.inputID,
+                shouldSaveDraft: props.shouldSaveDraft,
+                onBlur: props.onBlur,
+                onChange: (text) => {
```
Comment (luacmartins) [inline src/components/AddressSearch.js:162]: Maybe I'm wrong, but I believe that `onChanteText` passes the text value as argument vs `onChange` passes an event. We should make sure that we are passing the text value here.

**Review by luacmartins — CHANGES_REQUESTED** [review]
Left a few comments. I also see this error in Storybook when typing into the field (possibly related to my comment about `onChange`)

Screen Shot 2022-02-14 at 12 44 36 PM

File: src/components/AddressSearch.js (hunk 2 above)
Comment (kakajann) [reply src/components/AddressSearch.js:56]: Removed

File: src/components/AddressSearch.js (hunk 1 above)
Comment (kakajann) [reply src/components/AddressSearch.js:62]: Done

File: src/components/AddressSearch.js (hunk 3 above)
Comment (kakajann) [reply src/components/AddressSearch.js:145]: I think we should early return if there's no ref provided.

Just tried removing it and got this error:
Screen Shot 2022-02-15 at 6 59 18 PM

File: src/components/AddressSearch.js (hunk 4 above)
Comment (kakajann) [reply src/components/AddressSearch.js:162]: I think it returns only text value, considering `AddressSearch` is using `TextInput` as the base component.

https://github.com/Expensify/App/blob/c049cc4d16ec71606fb635a7d7913d21b0e876f1/src/components/TextInput/BaseTextInput.js#L240
https://github.com/Expensify/App/blob/c049cc4d16ec71606fb635a7d7913d21b0e876f1/src/components/TextInput/BaseTextInput.js#L121-L128

**Comment by kakajann** [comment]
> Left a few comments. I also see this error in Storybook when typing into the field (possibly related to my comment about `onChange`)
> 
> Screen Shot 2022-02-14 at 12 44 36 PM

Is this error occurs on the `Form` stories or the `AddressSearch`?

File: src/components/AddressSearch.js (hunk 4 above)
Comment (kakajann) [reply src/components/AddressSearch.js:161]: I don't think so.
Should I push this diff?
```diff
diff --git a/src/components/AddressSearch.js b/src/components/AddressSearch.js
index 005438f75..bee2e40cb 100644
--- a/src/components/AddressSearch.js
+++ b/src/components/AddressSearch.js
@@ -158,7 +158,7 @@ const AddressSearch = (props) => {
                 inputID: props.inputID,
                 shouldSaveDraft: props.shouldSaveDraft,
                 onBlur: props.onBlur,
-                onChange: (text) => {
+                onChangeText: (text) => {
                     if (skippedFirstOnChangeTextRef.current) {
                         props.onChange({street: text});
                     } else {
```

Comment (luacmartins) [reply src/components/AddressSearch.js:161]: @kakajann Sorry for the late reply, I was ooo for most of last week. I think it's fine to keep `onChangeText` as we had before. The `onChange` handler will be passed by the parent Form component.

Comment (kakajann) [reply src/components/AddressSearch.js]: @luacmartins, done

**Review by luacmartins — COMMENTED** [review]
Thanks for the changes @kakajann! I believe since `props.value` can be undefined, we should add a check [here](https://github.com/Expensify/App/blob/68ff8d070ba9ca8d2320236f6f44d1404ea9bed0/src/components/AddressSearch.js#L69) so we don't try to access `.trim()` on undefined.

**Comment by kakajann** [comment]
@luacmartins Done

File: src/components/AddressSearch.js [hunk 5]
```diff
@@ -9,13 +9,34 @@ import styles from '../styles/styles';
...
+    /** Indicates that the input is being used with the Form component */
+    isFormInput: PropTypes.bool,
+
+    /**
+     * The ID used to uniquely identify the input
+     *
+     * @param {Object} props - props passed to the input
+     * @returns {Object} - returns an Error object if isFormInput is supplied but inputID is falsey or not a string
+     */
+    inputID: props => FormUtils.getInputIDPropTypes(props),
```
Comment (luacmartins) [inline src/components/AddressSearch.js:29]: We renamed this util function to `validateInputIDProps`. Please update it.

Comment (luacmartins) [inline src/components/AddressSearch.js:29]: We renamed this util function to `validateInputIDProps`. Please update it.

**Review by luacmartins — COMMENTED** [review]
Sorry for all the back and forth. Left a few more comments.

File: src/components/AddressSearch.js (hunk 5 above)
Comment (kakajann) [reply src/components/AddressSearch.js:29]: Done

**Comment by kakajann** [comment]
No problem @luacmartins

**Review by luacmartins — APPROVED** [review]
LGTM! Thanks for all the changes @kakajann! @rushatgabhane and @marcochavezf all yours.

**Comment by rushatgabhane** [comment]
Thanks for the review @luacmartins 🙇

**Review by rushatgabhane — APPROVED** [review]
@kakajann could you please merge `main` into your branch.

@marcochavezf LGTM! 🎉️
Tests well, and I gave this a run after merging `main` from upstream.

**Comment by kakajann** [comment]
@rushatgabhane Done

**Review by marcochavezf — APPROVED** [review]

**Pull request merged**